File: algorithms/medium/minimum_swaps_to_group_all_1s_together.go
package main

import (
	"fmt"
    "math"
)

func minSwaps(data []int) int {
    size := 0
    for _, v := range data {
        if v == 1 {
            size++
        }
    }
    N := len(data)
    zeroCount := 0
    res := math.MaxInt
    for i := 0; i < N; i++ {
        if i < size {
            if data[i] == 0 {
                zeroCount++
            }
        } else {
            res = min(res, zeroCount)
            if data[i - size] == 0 {
                zeroCount--
            }
            if data[i] == 0 {
                zeroCount++
            }
        }
    }
    res = min(res, zeroCount)
    return res
}

func main() {
	lists := [][]int{
                 []int{1,0,1,0,1},
                 []int{0,0,0,1,0},
                 []int{1,0,1,0,1,0,0,1,1,0,1},
                 []int{0,0,1,1,1,1,0,0,1,1,1,1,0,1,0,1,1,0,1,0,0,1,0,1,0,0,0,0,0,1,1,0,1,1,1,1,1,1,0,1,0,0,1,0,1,1,0,0,0,0},
	}
	for _, data := range lists {
		fmt.Printf("data = %v\n", data)
	    r := minSwaps(data)
        fmt.Printf("r = %d\n", r)
		fmt.Println("===========================")
	}
}

File: algorithms/medium/minimum_swaps_to_group_all_1s_together.py
from typing import List

class Solution:
    def minSwaps(self, data: List[int]) -> int:
        # Intuition - Group the ones *in any place* in the array. The phrase "in any place"
        # is the giveaway. Use sliding window of size = freq_of_ones to see how many swaps
        # are needed.
        N = len(data)
        x = data.count(1)  # Frequency of 1s in the array
        zero_count = 0
        res = float('inf')
        for i in range(N):
            if data[i] == 0:
                zero_count += 1
            if i - x >= 0 and data[i - x] == 0:
                zero_count -= 1
            if i >= x - 1:
                res = min(res, zero_count)
        return res

# Main section
for data in [
               [1,0,1,0,1],
               [0,0,0,1,0],
               [1,0,1,0,1,0,0,1,1,0,1],
               [0,0,1,1,1,1,0,0,1,1,1,1,0,1,0,1,1,0,1,0,0,1,0,1,0,0,0,0,0,1,1,0,1,1,1,1,1,1,0,1,0,0,1,0,1,1,0,0,0,0],
            ]:
    print(f'data = {data}')
    sol = Solution()
    r = sol.minSwaps(data)
    print(f'r = {r}')
    print('========================')

